File: archive/2021s/solutions/FutureGadgetLab.py
from sys import stdin, stdout, setrecursionlimit
from math import gcd

n, j, k = (int(x) for x in stdin.readline().split())
edges = [[] for i in range(n)]
for i in range(n-1):
	u, v = (int(x) for x in stdin.readline().split())
	edges[u-1].append(v-1)


prob = [0]*n
g = 0
stack = [(0, 1, 0)]
def dfs(i):
	global g
	u, x, d = stack[i]
	prob[u] = x
	for child in edges[u]:
		stack.append((child, x/len(edges[u]), d+1))
	if len(edges[u]) == 0:
		g = gcd(g, d+1)

i = 0
while i < len(stack):
	dfs(i)
	i+=1
	
res = pow(j, k, g)
s = sum(prob)
ans = 0
stack = [(0, 0)]
def dfs2(i):
	global ans
	u, d = stack[i]
	if d%g == res:
		ans += prob[u]*g/s*d
	for child in edges[u]:
		stack.append((child, d+1))

i = 0
while i < len(stack):
	dfs2(i)
	i += 1

stdout.write(str(ans)+"\n")

File: archive/2021s/solutions/FutureGadgetLab.java
import java.util.*;
import java.io.*;

public class FutureGadgetLab {

	static ArrayList<Integer>[] adj;
	static Stack<double[]> stack;
	static double[] prob;
	static int g, res;
	static double ans, sum;

	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt(), j = sc.nextInt(), k = sc.nextInt();
		adj = new ArrayList[n];
		for (int i = 0; i < n; i++)
			adj[i] = new ArrayList<>();
		for (int i = 0; i < n - 1; i++) {
			int u = sc.nextInt();
			int v = sc.nextInt();
			adj[u - 1].add(v - 1);
		}
		prob = new double[n];
		g = 0;
		stack = new Stack<>();
		stack.add(new double[] { 0, 1, 0 });
		int i = 0;
		while (i < stack.size()) {
			dfs(i);
			i++;
		}
		res = (int) pow(j, k);
		sum = 0;
		for (double d : prob)
			sum += d;
		ans = 0;
		stack = new Stack<>();
		stack.add(new double[] { 0, 0 });
		i = 0;
		while (i < stack.size()) {
			dfs2(i);
			i++;
		}
		System.out.println(ans);
		sc.close();
	}

	static void dfs(int i) {
		double[] temp = stack.get(i);
		int u = (int) temp[0];
		double x = temp[1];
		int d = (int) temp[2];
		prob[u] = x;
		for (int v : adj[u])
			stack.add(new double[] { v, x / adj[u].size(), d + 1 });
		if (adj[u].size() == 0)
			g = gcd(g, d + 1);
	}

	static void dfs2(int i) {
		double[] temp = stack.get(i);
		int u = (int) temp[0];
		int d = (int) temp[1];
		if (d % g == res)
			ans += prob[u] * (double) g / (double) sum * d;
		for (int v : adj[u])
			stack.add(new double[] { v, d + 1 });
	}

	static long pow(long a, long b) {
		if (b == 0)
			return 1L;
		long val = pow(a, b / 2);
		if (b % 2 == 0)
			return val * val % g;
		else
			return val * val % g * a % g;
	}

	static int gcd(int a, int b) {
		if (a > b)
			return gcd(b, a);
		if (a == 0)
			return b;
		return gcd(b % a, a);
	}

}